### DISCUSSION TITLE
Performance questions

### DISCUSSION THREAD

**Discussion by mgberg** [oxigraph/oxigraph#1092] [General]
I'm interested in using pyoxigraph via oxrdflib to potentially improve performance of some tools I'm working on. I wrote a quick Python script to try a few different operations using RDFLib's default in-memory store, oxrdflib's in-memory store, and oxrdflib's persistent store.
For each:
- Parse data
- List all the classes (one usage of Graph.subjects/Store.quads_for_pattern)
- Do some RDF API operations (many usages of Graph.triples/Store.quads_for_pattern)
- Run a SPARQL query equivalent to the above
- Serialize data

Here's the results I saw:
```
TEST: RDFLib Memory
Parsed data in 59.257 ms
Found 84 classes via RDF APIs in 0.227 ms
Found 4339 terminal nodes via RDF APIs in 91.963 ms
Found 4339 terminal nodes via SPARQL query in 752.354 ms
Serialized data in 58.167 ms

TEST: Oxigraph In-Memory
Parsed data in 19.573 ms
Found 84 classes via RDF APIs in 0.782 ms
Found 4339 terminal nodes via RDF APIs in 240.453 ms
Found 4339 terminal nodes via SPARQL query in 19.992 ms
Serialized data in 5.556 ms

TEST: Oxigraph Persistent
Parsed data in 44.008 ms
Found 84 classes via RDF APIs in 1.029 ms
Found 4339 terminal nodes via RDF APIs in 299.757 ms
Found 4339 terminal nodes via SPARQL query in 37.503 ms
Serialized data in 7.482 ms
```

Two questions jump out at me from these results:
1. Oxigraph is _much_ faster than RDFLib for all operations except direct triple pattern matching. Is this primarily because of the number of times and the volume of data that is round-tripped from RDFLib to oxrdflib to rust and back? Is there anything that can be done (either on my end or in oxigraph) that could boost that performance?
2. Why does parsing data into an oxigraph persistent store take so much longer than parsing data into an oxigraph in-memory store? A bit longer is expected, but this was a bit surprising. I saw similar time discrepancies when using pyoxigraph directly instead of via RDFLib.

Thanks!

**Comment by Tpt** [comment]
Thank you for these numbers! They seems to be in the magnitude of what I expected.

Quick question: how are you doing data loading? Oxrdflib 0.4 exposes pyoxigraph parser and serializer that uses some hack to skip Python entirely when used (e.g. `graph.load(path="my_file.ttl", format="ox-ttl")`). 

>  Is this primarily because of the number of times and the volume of data that is round-tripped from RDFLib to oxrdflib to rust and back?

Yes! When doing triple pattern matching rdflib just does a lookup in its indexes and returns the existing rdflib triples whereas oxigraph also has to build Rust RDF terms, then pyoxigraph terms (e.g. `pyoxigraph.NamedNode`) and oxrdflib convert the terms to rdflib terms (e.g. `rdflib.URIRef`). That's quite a bit of conversion.

>  Is there anything that can be done (either on my end or in oxigraph) that could boost that performance?

On your end not much. In Oxigraph we can make lookup faster, optimize the way RDF terms are encoded and who the conversion is done in `pyoxigraph` and `oxrdflib`. For example [this succession of `isinstance` calls](https://github.com/oxigraph/oxrdflib/blob/bedbe715a5c8267dbe89b25863b99a5fffda831a/oxrdflib/_converter.py#L95) is maybe not the fastest but it's hard to optimize without good benchmarks.

> Why does parsing data into an oxigraph persistent store take so much longer than parsing data into an oxigraph in-memory store?

Two possible reasons:
- parsing into a persistent store means writing data to disk and this has some costs
- parsing without `transactional=False` option means loading atomicity is enabled and it has some cost (the in-memory storage also ensures atomicity but it's much easier in-memory)

**Reply by mgberg** [reply]
That was using the `ox-ttl` parser.

What `transactional=False` are you referring to?

**Reply by Tpt** [reply]
> That was using the ox-ttl parser.

Thank you!

> What transactional=False are you referring to?

This syntax: `store.load(path="my_file.ttl", format="ox-ttl", transactional=False)`. It disables Oxigraph atomicity.
